src/mapping/query/functions.rs

```diff
@@ -280,20 +280,53 @@ impl Function for UpcaseFn {
 
 #[derive(Debug)]
 pub(in crate::mapping) struct DowncaseFn {
-    query: Box<dyn Function>,
+    path: Box<dyn Function>,
+    append: Option<Box<dyn Function>>,
 }
 
 impl DowncaseFn {
-    pub(in crate::mapping) fn new(query: Box<dyn Function>) -> Self {
-        Self { query }
+    pub(in crate::mapping) fn new(
+        arguments: Vec<(Option<&str>, Box<dyn Function>)>,
+    ) -> Result<Self> {
+        let mut args = arguments
+            .into_iter()
+            .enumerate()
+            .map(|arg| match arg {
+                (0, (None, query)) => Ok(query),
+                (_, (Some("append"), query)) => Ok(query),
+
+                (i, (None, _)) => Err(format!("unexpected positional argument at {}", i)),
+                (0, (Some(_), _)) => Err("missing positional argument at 0".to_owned()),
+                (_, (Some(key), _)) => Err(format!("unexpected named argument: {}", key)),
+            })
+            .collect::<Result<Vec<_>>>()?;
+
+        let (append, path) = match args.len() {
+            0 => return Err("missing positional argument at 0".to_owned()),
+            1 => (None, args.remove(0)),
+            2 => (args.pop(), args.remove(0)),
+            _ => unreachable!(),
+        };
```

Comment:
This part validates the function signature the `downcase` function expects.

As you can see, it's quite cumbersome and error-prone I believe.

Some questions:

* Is this the general direction we want to take w.r.t. signature validation?
* Do we want to introduce some kind of `FunctionArg` type or a collection type to make it easier for functions to fetch their required positional/named arguments and error if something is missing?

Reply:
1. No, I don't think this is a good direction long term. Consolidating parsing in one place would make more sense to me. This would require #2 below.
2. Having each function type provide info about its arguments in a new type makes sense. Since we're handling it manually above, it doesn't have to be a trait, so it can be a `const ARGUMENTS: &'static [ArgumentDescriptor];` or whatever, as opposed to a function.
3. Definitely agree having a strict rule about which are positional and which are named makes this all much simpler.

Reply:
>     1. Consolidating parsing in one place would make more sense to me. This would require #2 below.

Yeah, that makes perfect sense, thanks for pointing me in that direction 👍

I'll investigate exposing signature details (using your `ArgumentDescriptor` example) of the function to the parser, so it can handle preparing the valid set of arguments to be sent to the function at runtime.

Do you happen to have any prior art w.r.t. such a system? I guess I could go read up on many of the existing languages, or some of the new (toy/scripting) languages built in Rust, to see how they do this.

Reply:
Sorry, no, while I have some awareness of scripting languages written in Rust, I have not investigated any of their internals.

Reply:
I think this would be a great place to use a procedural macro. We could implement a function like:

```rust
#[derive(Remap)]
struct SomeFunction {
  something: String,
  optional: Option<usize>,
  [name="thing"]
  namedArg: String
}

impl Function for SomeFunction {
...
```

The macro could generate the code to register itself and it's arguments with the parser (perhaps using `inventory`?). 

I think the parser would have to parse the args into a `Map<String, Box<dyn Function>>`. A wrapper struct could be generated that also implements Function. It would take the map, execute the params, validate that they evaluate to the correct types and then if they do, create `SomeFunction` and call `execute` on it.

Reply:
For absolute minimal boilerplate, it might actually be possible to create it as an attribute on a function:

```rust
#[remap named=(named_arg)]
fn some_function(something: String, optional: Option<usize>, named_arg: String) -> Value {
...
}
```

Reply:
I opted against using a procedural macro, as it only complicates things for little gain. I did introduce a macro_rules macro to make it easier to add new functions going forward.

With this change, the following changed from the _before_ situation:

* you **no longer** need to implement the `new` function
* you **no longer** need to add the function to the parser logic
* you **do** need to implement `TryFrom<ArgumentList>`
* you **do** need to add a single line to the macro, e.g. `"downcase" => DowncaseFn`
* you **do** need to implement the `Function::parameters()` method with returns a `&'static [Parameter]` list (unless the function takes no arguments, in which case the default impl suffices)

As an example, here's the `DowncaseFn` impl:

```rust
#[derive(Debug)]
pub(in crate::mapping) struct DowncaseFn {
    query: Box<dyn Function>,
}

impl Function for DowncaseFn {
    fn execute(&self, ctx: &Event) -> Result<Value> {
        match self.query.execute(ctx)? {
            Value::Bytes(bytes) => Ok(Value::Bytes(
                String::from_utf8_lossy(&bytes).to_lowercase().into(),
            )),
            value => unexpected_type!(value),
        }
    }

    fn parameters() -> &'static [Parameter] {
        &[Parameter {
            keyword: "value",
            kinds: &[ValueKind::Bytes],
            required: true,
        }]
    }
}

impl TryFrom<ArgumentList> for DowncaseFn {
    type Error = String;

    fn try_from(mut arguments: ArgumentList) -> Result<Self> {
        let query = arguments.required("value")?;

        Ok(Self { query })
    }
}
```